Add typing to ``filepath``

DanielNoord opened the pull request:
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

This fails the test suite, but I think this might be because the test suite is testing something that doesn't happen in normal use of pylint. Please correct me if I am wrong but after investigating the call-chain I believe that almost all `filepath`'s handled by `pylint` are `str` (the only exception being `get_fatal_error_message()`). This presumes that no user directly calls `linter.check()` as we do in the test suite. 
Of note is `lint.py_run()`, which I believe also does no run `check()` before calling `load_command_line_configuration`.

The following reasoning leads me to believe that a normal use of `pylint` always creates a `str` file path.
- `load_command_line_configuration()` in `pylint/config/option_manager_mixin.py` always returns a `List[str]`. Therefore the `args` variable in the `__init__` of the `Run` class in `pylint/lint/run.py` is always a `List[str]`.
- Subsequently in `pylint/lint/pylinter.py` the `files_or_modules` argument to `check()` is always a `List[str]`. Which means that both `check_parallel()` and `_check_files()` only handle `str` representations of paths.
- Therefore `set_current_module()` and `on_set_current_module()` in various other classes only get given `str`.

https://github.com/DanielNoord/pylint/pull/10
This PR shows that with changes to the tests this PR can succeed. These changes are:
- Making sure all initial file paths are a `List[str]`
- Making sure that `test_worker_check_single_file_uninitialised()` catches the new explicitly defined exception
- Making sure that `get_fatal_error_message()` always gets a `Path` as second argument
- Making sure that `check_parallel()` always gets an `Iterable`

These changes seem to follow the standard format of these arguments/variables in normal runs of `pylint` that are being skipped in the tests.

DanielNoord commented on pylint/lint/expand_modules.py:
```diff
@@ -1,9 +1,28 @@
+    from typing_extensions import Literal, TypedDict
+
+
+class MODULE_DESCRIPTION_DICT(TypedDict):
```
After a (potential) merge of #4973 these should be added to `pylint/typing`.
This PR is therefore blocked by #4973

cdce8p replied:
This should use `PascalCase`. Same for `ERROR_DESCRIPTION_DICT` below.

DanielNoord commented on pylint/lint/parallel.py:
```diff
@@ -62,9 +63,13 @@ def _worker_initialize(linter, arguments=None):
     name, filepath, modname = file_item
 
+    if not _worker_linter:
+        raise Exception("Worker linter not yet initialised")
```
`mpy` complains on the following two lines about the possibility that `_work_linter` might be None.
This is probably the wrong `Exception` type.
We currently already test for this, but do so with catching an `AttributeError` on L73. See:
https://github.com/PyCQA/pylint/blob/fab2de272b3fc36dc4480d89840b8bfe0fbfbd0c/tests/test_check_parallel.py#L177-L182
I think making this error more explicit as `mypy` wants makes sense.

cdce8p commented on pylint/lint/parallel.py:
```diff
@@ -62,9 +63,13 @@ def _worker_initialize(linter, arguments=None):
 
-def _worker_check_single_file(file_item):
+def _worker_check_single_file(
+    file_item: Tuple[str, str, str]
```
Does it make sense to define the type for `file_item` as NamedTuple? It seems to be used in a few places.

DanielNoord replied:
Good call!
Probably after https://github.com/PyCQA/pylint/pull/4973 though to add it to the `typing.py` document.

cdce8p reviewed (commented):
Not a review, just some thoughts.

cdce8p commented on pylint/lint/parallel.py:
```diff
@@ -151,7 +156,7 @@ def check_parallel(linter, jobs, files, arguments=None):
             for msg in messages:
                 msg = Message(*msg)
-                linter.reporter.handle_message(msg)
+                linter.reporter.handle_message(msg)  # type: ignore # L134 makes linter always have a reporter attribute
```
Mentioning line numbers in comments isn't worth it. The code will eventually get updated without the comment.

cdce8p commented on pylint/lint/parallel.py:
```diff
@@ -151,7 +156,7 @@ def check_parallel(linter, jobs, files, arguments=None):
             for msg in messages:
                 msg = Message(*msg)
-                linter.reporter.handle_message(msg)
+                linter.reporter.handle_message(msg)  # type: ignore # set_reporter() call above makes linter have a reporter attr
```
```suggestion
                linter.reporter.handle_message(msg)  # type: ignore  # linter.set_reporter() call above makes linter have a reporter attr
```

cdce8p reviewed (commented):
I wouldn't necessarily wait for #4973 to add `typing.py`. You could also add it here. It shouldn't be too hard to resolve the merge conflicts in the end.

For `Tuple[str, str, str]` it might even make sense to create a separate one. You should you by now that I like small PRs 😄

DanielNoord commented:
@cdce8p I added the NamedTuple. However, doing this separately was (quite) awful because of the various new `mypy` errors that I already fixed in this PR. (For example raising the Exception on `if not _worker_linter`).
So I added it here!

cdce8p commented on pylint/lint/expand_modules.py:
```diff
@@ -1,9 +1,28 @@
+    from typing_extensions import Literal, TypedDict
+
+
+class ModuleDescriptionDict(TypedDict):
+    path: str
+    name: str
+    isarg: bool
+    basepath: str
+    basename: str
+
+
+class ErrorDescriptionTest(TypedDict):
+    key: Literal["fatal"]
+    mod: str
+    ex: Union[ImportError, SyntaxError]
```
You can move these to `pylint/typing.py` now.

cdce8p commented on pylint/lint/parallel.py:
```diff
@@ -109,7 +115,7 @@ def _merge_mapreduce_data(linter, all_mapreduce_data):
 
 
-def check_parallel(linter, jobs, files, arguments=None):
+def check_parallel(linter, jobs, files: Iterator[FileItem], arguments=None):
```
Does it need to be an `Iterator` or is `Iterable` enough?
https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable

cdce8p commented on pylint/lint/pylinter.py:
```diff
@@ -935,16 +937,13 @@ def initialize(self):
                 self._msgs_state[msg.msgid] = False
 
-    def check(self, files_or_modules):
+    def check(self, files_or_modules: List[str]) -> None:
```
`Iterable`?

cdce8p commented on pylint/lint/pylinter.py:
```diff
@@ -970,7 +969,7 @@ def check(self, files_or_modules):
             )
 
-    def check_single_file(self, name, filepath, modname):
+    def check_single_file(self, name: str, filepath: str, modname: str) -> None:
```
These arguments seems to represent a `FileItem`.

DanielNoord replied:
Needed some refactoring, but added!

cdce8p commented on pylint/lint/pylinter.py:
```diff
@@ -982,7 +981,11 @@ def check_single_file(self, name, filepath, modname):
+    def _check_files(
+        self,
+        get_ast,
+        file_descrs: Union[List[FileItem], Iterator[FileItem]],
```
```suggestion
        file_descrs: Iterable[FileItem],
```

cdce8p commented on pylint/lint/pylinter.py:
```diff
@@ -982,7 +981,11 @@ def check_single_file(self, name, filepath, modname):
+        self,
+        get_ast,
+        file_descrs: Union[List[FileItem], Iterator[FileItem]],
+    ):
```
```suggestion
    ) -> None:
```

DanielNoord replied:
Done in own commit!

cdce8p commented on pylint/lint/pylinter.py:
```diff
@@ -1004,12 +1007,14 @@ def _check_files(self, get_ast, file_descrs):
 
-    def _check_file(self, get_ast, check_astroid_module, name, filepath, modname):
+    def _check_file(
+        self, get_ast, check_astroid_module, name: str, filepath: str, modname: str
```
These again seem to represent a `FileItem`.

cdce8p reviewed (commented):
I haven't commented on that before. A best practice when typing in python is to use the most generic type possible for arguments and the most concrete one for return types. I.e.
```py
def func(var: Iterable[str]) -> List[str]:
    ...
```

If it doesn't matter if a `list`, `tuple` or something else is pass in as long as it's possible to iterate over it, it shouldn't matter to the type checker either.

cdce8p commented on pylint/lint/parallel.py:
```diff
@@ -79,7 +83,7 @@ def _worker_check_single_file(file_item):
         id(multiprocessing.current_process()),
         _worker_linter.current_name,
-        filepath,
+        file_item[1],
```
```suggestion
        file_item.filepath,
```
One advantage of using a NamedTuple.

cdce8p commented on pylint/lint/pylinter.py:
```diff
@@ -935,16 +937,13 @@ def initialize(self):
                 self._msgs_state[msg.msgid] = False
 
-    def check(self, files_or_modules):
+    def check(self, files_or_modules: Sequence[str]) -> None:
```
`Sequence` is also an excellent choice!

cdce8p commented on pylint/lint/pylinter.py:
```diff
@@ -935,16 +937,13 @@ def initialize(self):
         """main checking entry: check a list of files or modules from their name.
 
-        files_or_modules is either a string or list of strings presenting modules to check.
+        files_or_modules is a list of strings presenting modules to check.
```
```suggestion
        files_or_modules is a sequence of strings presenting modules to check.
```

cdce8p commented on pylint/lint/pylinter.py:
```diff
@@ -1004,23 +1005,22 @@ def _check_files(self, get_ast, file_descrs):
-        :param str modname: name of the checked Python module
+        :param FileItem file: data about the file
         """
+        name, filepath, modname = file
```
Why do a tuple unpacking. With `NamedTuple` you can access the attributes by name.
It might even make sense to just pass `file` to `set_current_module` / `get_ast` and do the attribute access there. A bit more refactoring though.

DanielNoord replied:
I left the refactor for another time, since this is becoming quite a large PR.

One thing I noticed is some ambiguity between `modname` and `modpath`. They seem to be used for the same attribute. We currently use `modpath`, is that okay? Or do you want to start using `modname`?

Pierre-Sassoulas replied:
+1 for keeping the refactors for later :)

cdce8p reviewed (commented):
I really like where this is going!

Pierre-Sassoulas commented:
>  I believe that almost all filepath's handled by pylint are str

We should handle ``Union[Path, str]`` imo.

DanielNoord commented:
> We should handle `Union[Path, str]` imo.

Not sure if I agree here. This would probably lead us to having to add `Path` to many other typings down the line. The current flow of `pylint` clearly does not allow `Path` to be passed. With a little refactor we can even get the last instance of `Path` to also become a `str`. Having to add `Union[Path, str]` instead of `str` everywhere might clutter the codebase.

@cdce8p & @Pierre-Sassoulas Do you guys thinks your review is far enough for me to add the commit with test fixes as discussed in the PR description and shown in https://github.com/DanielNoord/pylint/pull/10?

Pierre-Sassoulas commented:
> might clutter the codebase.

Yes, also the migration to Path is also a huge endeavor with not a lot of clear benefit right now. Most ``os.path`` handles ``Union[Path, str]`` so technically that's what we can probably handle on most of the function, but finding out where we can't is probably not worth the time and crashes it will create.

DanielNoord commented:
This now includes the relevant changes to the tests to make this work. As discussed in PR description I believe the tests check things that cannot happen in real world scenarios. The description includes some additional reasoning for this and a description of the changes to the tests.

Pierre-Sassoulas commented on pylint/lint/pylinter.py:
```diff
@@ -938,16 +939,13 @@ def initialize(self):
         """
         self.initialize()
 
-        if not isinstance(files_or_modules, (list, tuple)):
-            files_or_modules = (files_or_modules,)
-
```
I like this, clearer input, it will make the code better. But I think this might be a breaking change.

cdce8p replied:
Technically all changes of function arguments are breaking ones as well (`check_single_file`, `check`, `_check_files`). It depends if we would consider them part of the public API. Something I'm not sure about.

Pierre-Sassoulas replied:
Well I think the check function can be used by addon relying on pylint's checker. I'm for changing the typing only right now and removing the mechanism in 3.0 as a breaking change (we can also add a deprecation warning).

Pierre-Sassoulas commented on pylint/lint/pylinter.py:
```diff
@@ -1059,19 +1057,19 @@ def _get_file_descr_from_stdin(filepath):
             modname = os.path.splitext(os.path.basename(filepath))[0]
 
-        return (modname, filepath, filepath)
+        return FileItem(modname, filepath, filepath)
```
👍

Pierre-Sassoulas commented on pylint/lint/pylinter.py:
```diff
@@ -938,16 +939,13 @@ def initialize(self):
+        files_or_modules is a sequence of strings presenting modules to check.
         """
         self.initialize()
 
```
```suggestion
        if not isinstance(files_or_modules, (list, tuple)):
            warnins.warn("In pylint 3.0, the checkers check function will only accept sequence of string")
            files_or_modules = (files_or_modules,)
```

cdce8p commented on pylint/lint/pylinter.py:
```diff
@@ -938,16 +939,17 @@ def initialize(self):
                 self._msgs_state[msg.msgid] = False
 
-    def check(self, files_or_modules):
+    def check(self, files_or_modules: Sequence[str]) -> None:
```
`Union[Sequence[str], str]` maybe with a `TODO` comment after the deprecation warning to update the type annotation.

DanielNoord commented on pylint/lint/pylinter.py:
```diff
@@ -1007,32 +1011,30 @@ def _check_files(self, get_ast, file_descrs):
+                        self.add_message(symbol, args=msg)
 
-    def _check_file(self, get_ast, check_astroid_module, name, filepath, modname):
+    def _check_file(self, get_ast, check_astroid_module, file: FileItem):
```
As in, keep the `FileItem` or revert to previous state?

cdce8p replied:
😅 Keep the change with `FileItem`

cdce8p commented on pylint/lint/pylinter.py:
```diff
@@ -938,16 +939,18 @@ def initialize(self):
+    def check(self, files_or_modules: Union[Sequence[str], str]) -> None:
         """main checking entry: check a list of files or modules from their name.
 
-        files_or_modules is either a string or list of strings presenting modules to check.
+        files_or_modules is a sequence of strings presenting modules to check.
```
This should be reverted

DanielNoord replied:
Weird how Github doesn't allow a suggestion here..

cdce8p replied:
I've highlighted both an old an a new line. That's probably why.

cdce8p commented on pylint/lint/pylinter.py:
```diff
@@ -938,16 +939,18 @@ def initialize(self):
         if not isinstance(files_or_modules, (list, tuple)):
-            files_or_modules = (files_or_modules,)
-
+            warnings.warn(
```
```suggestion
            # pylint: disable-next=fixme
            # TODO: Update typing and docstring for 'files_or_modules' when removing the deprecation
            warnings.warn(
```

cdce8p commented on pylint/lint/pylinter.py:
```diff
@@ -938,16 +939,18 @@ def initialize(self):
-
+            warnings.warn(
+                "In pylint 3.0, the checkers check function will only accept sequence of string",
+                DeprecationWarning,  # TODO: Update typing of files_or_modules to Sequence[str] # pylint: disable=fixme
```
```suggestion
                DeprecationWarning,
```

cdce8p commented on pylint/lint/pylinter.py:
```diff
@@ -973,66 +976,64 @@ def check(self, files_or_modules):
-        """Check single file
+    def check_single_file(self, name: str, filepath: str, modname: str) -> None:
+        warnings.warn(
+            "In pylint 3.0, the checkers check_single_file function will only accept a FileItem",
```
```suggestion
            "In pylint 3.0, the checkers check_single_file function will be removed. "
            "Use check_single_file_item instead.",
```
What do you think?

DanielNoord replied:
Don't really have a preference. Guess this works as well and requires less refactoring for `3.0`.

cdce8p commented on pylint/lint/pylinter.py:
```diff
@@ -973,66 +976,64 @@ def check(self, files_or_modules):
+    def check_single_file(self, name: str, filepath: str, modname: str) -> None:
+        warnings.warn(
+            "In pylint 3.0, the checkers check_single_file function will only accept a FileItem",
+            DeprecationWarning,  # TODO: Replace check_single_file with check_single_file_item # pylint: disable=fixme
```
```suggestion
            DeprecationWarning,
```

cdce8p commented on pylint/typing.py:
```diff
@@ -2,11 +2,48 @@
+    Each file item has:
+    - name: full name of the module
+    - filepath: path of the file
+    - modname: module name"""
```
```suggestion
    - modname: module name
    """
```
PEP8 -> closing quotes should be on a new line if docstring > 1 line.
